include/exadg/poisson/spatial_discretization/operator.cpp

```diff
@@ -76,20 +77,52 @@ Operator<dim, Number, n_components>::distribute_dofs()
   if(n_components == 1)
   {
     if(param.spatial_discretization == SpatialDiscretization::DG)
-      fe = std::make_shared<FE_DGQ<dim>>(param.degree);
+    {
+      if(this->grid->triangulation->all_reference_cells_are_hyper_cube())
+        fe = std::make_shared<FE_DGQ<dim>>(param.degree);
+      else if(this->grid->triangulation->all_reference_cells_are_simplex())
+        fe = std::make_shared<FE_SimplexDGP<dim>>(param.degree);
+      else
+        AssertThrow(false, ExcNotImplemented());
+    }
     else if(param.spatial_discretization == SpatialDiscretization::CG)
-      fe = std::make_shared<FE_Q<dim>>(param.degree);
+    {
+      if(this->grid->triangulation->all_reference_cells_are_hyper_cube())
+        fe = std::make_shared<FE_Q<dim>>(param.degree);
+      else if(this->grid->triangulation->all_reference_cells_are_simplex())
+        fe = std::make_shared<FE_SimplexP<dim>>(param.degree);
+      else
+        AssertThrow(false, ExcNotImplemented());
+    }
     else
+    {
       AssertThrow(false, ExcMessage("not implemented."));
+    }
   }
   else if(n_components == dim)
   {
     if(param.spatial_discretization == SpatialDiscretization::DG)
-      fe = std::make_shared<FESystem<dim>>(FE_DGQ<dim>(param.degree), dim);
+    {
+      if(this->grid->triangulation->all_reference_cells_are_hyper_cube())
+        fe = std::make_shared<FESystem<dim>>(FE_DGQ<dim>(param.degree), dim);
+      else if(this->grid->triangulation->all_reference_cells_are_simplex())
+        fe = std::make_shared<FESystem<dim>>(FE_SimplexDGP<dim>(param.degree), dim);
+      else
+        AssertThrow(false, ExcNotImplemented());
+    }
     else if(param.spatial_discretization == SpatialDiscretization::CG)
-      fe = std::make_shared<FESystem<dim>>(FE_Q<dim>(param.degree), dim);
+    {
+      if(this->grid->triangulation->all_reference_cells_are_hyper_cube())
+        fe = std::make_shared<FESystem<dim>>(FE_Q<dim>(param.degree), dim);
+      else if(this->grid->triangulation->all_reference_cells_are_simplex())
```

Comment:
Just move the content of https://github.com/dealii/dealii/blob/be72dbb5a5f97aee1c8cf330b66664c848176fba/source/grid/tria.cc#L14668-L14669 into a free function: `bool all_reference_cells_are_simplex(Triangulation const & tria);`. You get `reference_cells` by calling `Triangulation::get_reference_cells()`.